tests/tls_e2e/common/openssl_utility.cpp

```diff
@@ -0,0 +1,315 @@
+        &private_key_buf_size);
+
+    OE_CHECK_MSG(result, " failed with %s\n", oe_result_str(result));
+    OE_TRACE_INFO("public_key_buf_size:[%ld]\n", public_key_buf_size);
+    OE_TRACE_INFO("public key used:\n[%s]", public_key_buf);
+
+    result = oe_generate_attestation_certificate(
+        (const unsigned char*)"CN=Open Enclave SDK,O=OESDK TLS,C=US",
+        private_key_buf,
+        private_key_buf_size,
+        public_key_buf,
+        public_key_buf_size,
+        &output_cert,
+        &output_cert_size);
+    OE_CHECK_MSG(result, " failed with %s\n", oe_result_str(result));
+
+    // temporary buffer required as if d2i_x509 call is successful cert_buf_ptr
+    // is incremented to the byte following the parsed data. sending
+    // cert_buf_ptr as argument will keep output_cert pointer undisturbed.
+    cert_buf_ptr = output_cert;
+
+    if ((cert = d2i_X509(NULL, &cert_buf_ptr, (long)output_cert_size)) == NULL)
+    {
+        OE_TRACE_ERROR(
+            TLS_SERVER
+            "Failed to convert DER fromat certificate to X509 structure\n");
+        goto done;
+    }
+    mem = BIO_new_mem_buf((void*)private_key_buf, -1);
+    if ((pkey = PEM_read_bio_PrivateKey(mem, NULL, 0, NULL)) == NULL)
+    {
+        OE_TRACE_ERROR(
+            TLS_SERVER
+            "Failed to convert private key buffer into EVP_KEY format\n");
+        goto done;
+    }
+
+    result = OE_OK;
+done:
+    cert_buf_ptr = NULL;
+    BIO_free(mem);
+    oe_free_key(private_key_buf, private_key_buf_size, NULL, 0);
+    oe_free_key(public_key_buf, public_key_buf_size, NULL, 0);
+    oe_free_attestation_certificate(output_cert);
+    return result;
+}
+
+// The return value of verify_callback controls the strategy of the further
+// verification process. If verify_callback returns 0, the verification process
+// is immediately stopped with "verification failed" state and A verification
+// failure alert is sent to the peer and the TLS/SSL handshake is terminated. If
+// verify_callback returns 1, the verification process is continued.
+int cert_verify_callback(int preverify_ok, X509_STORE_CTX* ctx)
+{
+    int ret = 0;
+    size_t der_len = 0;
+    unsigned char* der = nullptr;
+    unsigned char* buff = nullptr;
+    oe_result_t result = OE_FAILURE;
+    X509* crt = nullptr;
+    int err = X509_V_ERR_UNSPECIFIED;
+
+    if (g_control_config.fail_cert_verify_callback)
+    {
+        OE_TRACE_INFO(
+            "Purposely returns failure from server's cert_verify_callback()\n");
+        goto done;
+    }
+
+    OE_TRACE_INFO(
+        TLS_SERVER "verify_callback called with preverify_ok=%d\n",
+        preverify_ok);
+    crt = X509_STORE_CTX_get_current_cert(ctx);
+    if (crt == nullptr)
+    {
+        OE_TRACE_ERROR(TLS_SERVER "failed to retrieve certificate\n");
+        goto done;
+    }
+
+    if (preverify_ok == 0)
+    {
+        err = X509_STORE_CTX_get_error(ctx);
+        if (err == X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT)
```

Comment:
I'm not followed. Can you elaborate more on this behavior (when will this case occur)?

Reply:
This is picked from the existing code https://github.com/openenclave/openenclave/blob/master/samples/attested_tls/common/verify_callback.cpp#L210. From the log statements, I could understand that if the generated certificate is a self-signed one, we are catching that error and sending return value as 1, which keeps preverify_ok = 1 and further certificate verification is done in the next call. In total, cert_verify_callback would be called twice once with preverify_ok = 0 and second time with preverify_ok=1.  Adding @CodeMonkeyLeet to understand more details.

Reply:
It would at least be helpful to document this behavior here.